File: Lecture/Binary Trees/Assignment/level-wise-linkedlist.cpp
/***
 * Given a binary tree, write code to create a separate linked list for each level. You need to return the array which contains head of each level linked list.
 ***/
/**********************************************************
	Following are the Binary Tree Node class structure and
	the Node class structure

	template <typename T>
	class BinaryTreeNode {
    	public: 
    	T data;
    	BinaryTreeNode<T> *left;
    	BinaryTreeNode<T> *right;

    	BinaryTreeNode(T data) {
        	this->data = data;
        	left = NULL;
        	right = NULL;
    	}
	};

	template <typename T>
	class Node {
		public:
    	T data;
    	Node<T> *next;
    	Node(T data) {
        	this->data=data;
        	this->next=NULL;
    	}
	};

***********************************************************/
vector<Node<int>*> constructLinkedListForEachLevel(BinaryTreeNode<int> *root) {  
    vector<Node<int>*> v;
    if(root == NULL)
        return v;
    
    queue<BinaryTreeNode<int>*> q;
    q.push(root);
    q.push(NULL);
    
    Node<int> *head = NULL;
    Node<int> *tail = NULL;
    while(!q.empty()) {
        BinaryTreeNode<int> *front = q.front();
        q.pop();
        
        if(front == NULL) {
            v.push_back(head);
            head = NULL;
            tail = NULL;
            
            if(q.empty())
                break;
            q.push(NULL);
            continue;
        }
        
    	Node<int> *node = new Node<int>(front->data);
        if(head == NULL) {
            head = node;
            tail = node;
        }
        else {
            tail->next = node;
            tail = tail->next;
        }
        
        if(front->left != NULL)
            q.push(front->left);
        if(front->right != NULL)
            q.push(front->right);
    }
    return v;
}

File: Lecture/Binary Trees/Assignment/LevelWiseLinkedList.java
/***
 * Given a binary tree, write code to create a separate linked list for each level. You need to return the array which contains head of each level linked list.
 ***/
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/*
Time complexity: O(N)
Space complexity: O(H)
where N is the number of nodes in the input BST
and H is the height of the input BST
*/
public class LevelWiseLinkedList {
    public static ArrayList<LinkedListNode<Integer>> constructLinkedListForEachLevel(BinaryTreeNode<Integer> root) {
        if (root == null) {
            return null;
        }
        Queue<BinaryTreeNode<Integer>> pendingNodes = new LinkedList<BinaryTreeNode<Integer>>();
        pendingNodes.add(root);
        int currentLevelRemaining = 1;
        int nextLevelCount = 0;
        LinkedListNode<Integer> currentLevelHead = null;
        LinkedListNode<Integer> currentLevelTail = null;
        ArrayList<LinkedListNode<Integer>> output = new ArrayList<>();
        while (!pendingNodes.isEmpty()) {
            BinaryTreeNode<Integer> currentNode;
            currentNode = pendingNodes.remove();
            LinkedListNode<Integer> newNode = new LinkedListNode<Integer>(currentNode.data);
            if (currentLevelHead == null) {
                currentLevelHead = newNode;
                currentLevelTail = newNode;
            } else {
                currentLevelTail.next = newNode;
                currentLevelTail = newNode;
            }
            if (currentNode.left != null) {
                pendingNodes.add(currentNode.left);
                nextLevelCount++;
            }
            if (currentNode.right != null) {
                pendingNodes.add(currentNode.right);
                nextLevelCount++;
            }
            currentLevelRemaining--;
            if (currentLevelRemaining == 0) {
                output.add(currentLevelHead);
                currentLevelHead = null;
                currentLevelTail = null;
                currentLevelRemaining = nextLevelCount;
                nextLevelCount = 0;
            }
        }
        return output;
    }
}